### PR TITLE
FakeTimeProvider usage update

### LINKED ISSUE #4912: FakeTimeProvider.Advance fails to advance delays (at least in the case of Polly)
### Description

When using FakeTimerProvider in unit tests, retries configured in a ResiliencyPipeline are not executed.

The library maintainer believes the issue lies in FakeTimerProvider
https://github.com/App-vNext/Polly/issues/1932

### Reproduction Steps

Please see the minimal unit test repo
https://github.com/rcollette/PollyRetryFakeTimeProviderIssue

### Expected behavior

Delays used in Poly should be advanced.

### Actual behavior

They don't appear to be triggered.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 8.0.101
Polly 8.2.1

### Other information

I tried to debug down into the WakeWaiters/InvokeCallback/TimerCallback code but was not seeing why this was not triggering.

### PR DESCRIPTION
**Pull request by dariusclay**
Closes #4912 

Adds additional test and information regarding usage of FakeTimeProvider

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/extensions/pull/5169)

### CODE REVIEW

**Comment by dariusclay** [comment]
In the original thread #4912, the minimal repro overlooked the need to also advance due to the simulated task delay.

```cs
    [Fact]
    public void PollyRetry_WhenTaskDelayLessThanCancellationAndAboveRetryDelay_ShouldHave2Tries()
    {
        // Act
        var result = _someService.PollyRetry(taskDelay: 1, cancellationSeconds: 6);

        // Advance for simulated task delay
        _timeProvider.Advance(TimeSpan.FromMilliseconds(1000));

        // Advancing the time should resolve the first execution delay.
        _timeProvider.Advance(TimeSpan.FromMilliseconds(1000));

        // Advance for simulated task delay
        _timeProvider.Advance(TimeSpan.FromMilliseconds(1000));

        // Advancing the time should resolve the second execution delay.
        _timeProvider.Advance(TimeSpan.FromMilliseconds(1000));

        // Assert
        result.IsCompleted.Should().BeFalse();
        _someService.Tries.Should().Be(2);
    }
```

**Comment by dariusclay** [comment]
The root issue here is with the way `Task.Delay(..., timeProvider)` works under the hood. In the case for non-system time providers, it calls `provider.CreateTimer()` to schedule a timer which FakeTimeProvider adds to it's list of Waiters. These get scheduled to run outside of the current context and may not all get a chance to be fired when Advance is called.

One thing of interest to note here, is that using `Delay(...).ConfigureAwait(true)` works in unit tests even without sleeping in the Advance method, as it ensures the continuation occurs on the same execution context. This would put the burden on any library or user code to know that it is using the fake provider instead of the TimeProvider.System instance.

**Review by geeknoid — APPROVED** [review]

**Review by xakep139 — COMMENTED** [review]
LGTM

File: test/Libraries/Microsoft.Extensions.TimeProvider.Testing.Tests/FakeTimeProviderTests.cs [hunk 1]
```diff
@@ -381,4 +381,57 @@ public void ShouldResetGateUnderLock_PreventingContextSwitching_AffectionOnTimer
...
     }
+
+    [Fact]
+    public void SimulateRetryPolicy()
+    {
+        // Arrange
+        var retries = 42;
+        var tries = 0;
+        var taskDelay = 0.5;
+        var delay = 1;
```
Comment (RussKie) [inline test/Libraries/Microsoft.Extensions.TimeProvider.Testing.Tests/FakeTimeProviderTests.cs:392]: Please don't use `var` for intrinsic types.

File: test/Libraries/Microsoft.Extensions.TimeProvider.Testing.Tests/FakeTimeProviderTests.cs [hunk 2]
```diff
@@ -381,4 +381,57 @@ public void ShouldResetGateUnderLock_PreventingContextSwitching_AffectionOnTimer
...
+                catch (InvalidOperationException)
+                {
+                    // ConfigureAwait(true) is required to ensure that tasks continue on the captured context
+                    await provider.Delay(TimeSpan.FromSeconds(delay)).ConfigureAwait(true);
+                }
+            }
+        }
+
+        // Act
+        var result = simulatedPollyRetry();
```
Comment (RussKie) [inline test/Libraries/Microsoft.Extensions.TimeProvider.Testing.Tests/FakeTimeProviderTests.cs:422]: Please don't use `var` unless the type is obvious.
https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md #10

**Pull request merged**